data/lib/core/creature.lua

```diff
@@ -1,29 +1,25 @@
-function Creature.getClosestFreePosition(self, position, extended)
-	local usePosition = Position(position)
-	local tiles = { Tile(usePosition) }
-	local length = extended and 2 or 1
+function Creature.getClosestFreePosition(self, position, maxRadius, mustBeReachable)
+	maxRadius = maxRadius or 1
 
-	local tile
-	for y = -length, length do
-		for x = -length, length do
-			if x ~= 0 or y ~= 0 then
-				usePosition.x = position.x + x
-				usePosition.y = position.y + y
+	-- backward compatability (extended)
+	if maxRadius == true then
+		maxRadius = 2
+	end
 
-				tile = Tile(usePosition)
-				if tile then
-					tiles[#tiles + 1] = tile
+	for radius = 0, maxRadius do
+		for x = -radius, radius do
+			for y = -radius, radius do
+				if math.abs(x) == radius or math.abs(y) == radius then
+					local checkPosition = Position(position.x + x, position.y + y, position.z)
+					local tile = Tile(checkPosition)
+					if tile:getCreatureCount() == 0 and not tile:hasProperty(CONST_PROP_IMMOVABLEBLOCKSOLID) and
+						(not mustBeReachable or self:getPathTo(checkPosition)) then
+						return checkPosition
```

Comment:
Yeah, I'm more used to C++ than Lua and didn't look further. I'm not sure about function call overhead on Lua, I program in Python (where the overhead is huge) and JS (where the overhead is not significant, and indeed better than looping) so it should be analyzed too.